### PR TITLE
Peripheral discovery as async operations

### PR DESCRIPTION
**Pull request by Supereg**
# Peripheral discovery as async operations

## :recycle: Current situation & Problem
While SpeziBluetooth exposes all Peripheral interactions as async/await-based operations, embracing Swift Concurrency, it doesn't always do that for internal operations. For example, the initial discovering phase does not use async/await and instead performs checks within the delegate method implementations. This has the result, that code is relatively hard to read as state checking is scattered around the implementation (e.g., we check in several places if we can call `signalFullyDiscovered()`).
This PR improves code readability and state management by making all peripheral interactions relying on async/await-based operations. Operations like service and characteristic discovery can now be performed via an async call. Further, enabling notifications can be performed using an async method call. As a result, all code relating to service discovery is now placed in one async method. Potential errors can be easily propagated back to the `connect` continuation making it possible to properly communicate if a connecting attempt failed (e.g., because we weren't able to subscribe to a characteristic after several tries).

Further, this PR changes the point in time when we consider a device fully connected. While we are still subscribing to characteristic notifications, the device will still report `connecting` state. This makes it easier for library users to determine when a device is fully ready.

## :gear: Release Notes 
* Internal restructure of BluetoothPeripheral resulting in better maintainable code.
* Errors from the service discovery are now reported back to the `connect()` method call and errors are thrown into the continuation. An unsuccessful discovery will result in the device being disconnected.
* Device is considered `connecting` while characteristic notifications are enabled.
* Fixed an issue were the onChange handler would not be called if the `initial` parameter was set to `false` if the characteristic is notify-only or an error occurred in the initial read.
* `DiscoveryCriteria` now supports specifying multiple advertised services.
* Fixed a potential race condition when enabling notifications for a `@Characteristic` before the peripheral was injected.
* `CurrentTimeService/synchronizeDeviceTime(now:threshold:)` is now `async throws` allowing to inspect potential errors.

## :books: Documentation
Documentation was updated to reflect changes in this PR.

## :white_check_mark: Testing
Existing testing infrastructure was used to validate correctness.

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: Package.swift [hunk 1]
```diff
@@ -32,8 +32,8 @@ let package = Package(
         .library(name: "SpeziBluetooth", targets: ["SpeziBluetooth"])
     ],
     dependencies: [
-        .package(url: "https://github.com/StanfordSpezi/SpeziFoundation", from: "1.1.0"),
-        .package(url: "https://github.com/StanfordSpezi/Spezi", from: "1.6.0"),
+        .package(url: "https://github.com/StanfordSpezi/SpeziFoundation", from: "2.0.0-beta.1"),
+        .package(url: "https://github.com/StanfordSpezi/Spezi", branch: "feature/upgrade-spezi"),
```
Comment (PSchmiedmayer) [inline Package.swift:36]: TODO to fix this to a version once released.

File: Package.swift
Comment (PSchmiedmayer) [inline Package.swift:1]: We should also update the SwiftLint version number here.

**Review by PSchmiedmayer — COMMENTED** [review]
Thank you for working on this @Supereg; let me know once you need a detailed review of this PR 👍

**Comment by Supereg** [comment]
@PSchmiedmayer fyi, This ist the last PR planned for the 3.0 release. Once it is merged it would aim to push a release version of 3.0.

**Review by PSchmiedmayer — APPROVED** [review]
@Supereg Thank you for all the other tests; I have also reconfigured the runner and builds should pass now.

**Comment by Supereg** [comment]
> @Supereg Thank you for all the other tests; I have also reconfigured the runner and builds should pass now.

great, thank you 🚀

**Pull request merged**